src/Pathy/Path.purs

```diff
@@ -0,0 +1,291 @@
+
+-- | Creates a path which points to a relative directory of the specified name.
+dir' :: Name Dir -> Path Rel Dir
+dir' = In currentDir
+
+-- | Creates a path that points to the parent directory of the specified path.
+parentOf :: forall a. Path a Dir -> Path a Dir
+parentOf p = ParentOf p
+
+-- | Extends a path with a file or directory under the current path.
+extendPath :: forall a b. Path a Dir -> Name b -> Path a b
+extendPath p = In p
+
+-- | Given a directory path, appends a relative path to extend the original
+-- | path.
+appendPath :: forall a b. Path a Dir -> Path Rel b -> Path a b
+appendPath = case _, _ of
+  Init, Init -> Init
+  ParentOf p, Init -> ParentOf (p </> Init)
+  In p (Name d), Init -> In (p </> Init) (Name d)
+  p1, ParentOf p2 -> ParentOf (p1 </> p2)
+  p1, In p2 n -> In (p1 </> p2) n
+
+infixl 6 appendPath as </>
+
+-- | Ascends into the parent of the specified directory, then descends into
+-- | the specified path.
+-- |
+-- | ```purescript
+-- | canonicalize (rootDir </> dir "foo" <..> dir "bar") = rootDir </> dir "bar"
+-- | ```
+parentAppend :: forall a b. Path a Dir -> Path Rel b -> Path a b
+parentAppend d p = parentOf d </> p
+
+infixl 6 parentAppend as <..>
+
+-- | Canonicalizes a path, by reducing things in the form `/x/../` to just
+-- | `/x/`. Paths like `/../` will be normalized to `/`.
+canonicalize :: forall a b. IsRelOrAbs a => Path a b -> Path a b
+canonicalize p = fromMaybe p (go p)
+  where
+    go :: forall b'. Path a b' -> Maybe (Path a b')
+    go = case _ of
+      Init ->
+        Nothing
+      p'@(ParentOf Init) ->
+        foldRelOrAbs (const Nothing) (const (Just Init)) p'
+      ParentOf (In p' _) ->
+        -- Coercion is safe as `ParentOf` can only appear where `b' ~ Dir`
+        Just $ (unsafeCoerce :: Path a Dir -> Path a b') (canonicalize p')
+      ParentOf p' ->
+        case go p' of
+          Just p'' -> Just $ canonicalize (ParentOf p'')
+          Nothing -> Nothing
+      In p' n ->
+        flip In n <$> go p'
+
+-- | A fold over `Path`s. Since `Path` has private constructors, this allows for
+-- | functions to be written over its constructors, similar to a total pattern
+-- | match.
+-- |
+-- | - The first argument is the value to return for the `currentDir`/`rootDir`
+-- |   at the base of the path.
+-- | - The second argument is a function for handling a step into the parent
+-- |   directory of the path it receives (eliminates `parentOf`).
+-- | - The third argument is a function representing a file or directory within
+-- |   the directory of the path it receives (eliminates `extendPath`).
+foldPath
+  :: forall a b r
+   . r
+  -> (Path a Dir -> r)
+  -> (Path a Dir -> Name b -> r)
+  -> Path a b
+  -> r
+foldPath r f g = case _ of
+  Init -> r
+  ParentOf d -> f d
+  In d n -> g d n
+
+-- | Peels off the last directory and the terminal file or directory name
+-- | from the path. Returns `Nothing` if the path is `rootDir` / `currentDir` or
+-- | some `parentOf p`.
```

Comment:
Yeah, maybe. It is a little different than it used to be though, as previously if you had `/foo/../` it would peel `foo` out of there for you, whereas now you get `Nothing` - that's what I was trying to explain. (That seemed confusing to me, and if you really want it you can still do the old behaviour with `foldPath` now).

Reply:
Current peel will also return Nothing for `/foo/bar/../` when it could be `/foo x "bar"`. is it intended result? maybe we should canonicalize and then fold as it's now? or:
``` purs
peel = canonicalize >>> peel'
peel' = foldPath Nothing (const Nothing) (\p n -> Just (Tuple p n)) 
```
and export both

Reply:
✅  added `peel' = canonicalize >>> peel`